### ISSUE TITLE
DMatrix handling of one-hot labels (Python)

### ISSUE THREAD

**Issue by loewenm** [dmlc/xgboost#10095]
It appears that DMatrix does not handle one-hot encoded **labels** appropriately.

**Symptom:**

DMatrix flattens one-hot encoded labels into a 1d array for shape [samples * classes,] instead of preserving original shape.

**Example:**
```
from sklearn.datasets import make_multilabel_classification

X, y = make_multilabel_classification(
    n_samples=100, n_classes=5, n_labels=3, random_state=0
)
print("X: {} | y: {}".format(X.shape, y.shape))

>>> X: (100, 20) | y: (100, 5)
```
```
dmatrix_dataset = xgb.DMatrix(
    X,
    label=y,
)
print(
    "data: {} | labels: {}".format(
        dmatrix_dataset.get_data().shape, dmatrix_dataset.get_label().shape
    )
)

>>> data: (100, 20) | labels: (500,)
```

**Explanation of output:**

The original **y** variable in the above example has shape [100, 5] (e.g. five one-hot encoded labels)

However, if I try to extract labels from the DMatrix, it has been reshaped to [100 * 5, ].

**Question:**

Is this working as intended? Does DMatrix not support one-hot encoded labels?

**Additional Notes:**

There does not appear to be any parameter to fix this issue in the [official docs](https://xgboost.readthedocs.io/en/stable/python/python_api.html#module-xgboost.core).

**Comment by trivialfis** [comment]
At the moment, the DMatrix supports consuming 2-D labels not doesn't support returning them. We implemented the receiving part for DMatrix so that we can get basic multi-target/label training to work, the returning part is still working in progress as we are trying to improve the support for custom-objective with multi-target/label.

**Comment by trivialfis** [comment]
Closing in favor of https://github.com/dmlc/xgboost/issues/9043
